### PR TITLE
feature/21 - Masonite Scheduling

### PR DESCRIPTION
**Pull request by josephmancuso**
Merged the Masonite-scheduler package into this codebase. Actually only took literally 5 minutes.

We will abandon that masonite-scheduler repo https://github.com/MasoniteFramework/scheduler when Masonite 4 comes out.

This PR still needs:

- [x] Clean codebase and improve upon from original task scheduling codebase
- [ ] Look at any additional features this could support

It would be cool if we could run a lambda function. Something like this:

```python
self.schedule(lambda: (
    self.application.make('mail').mailable(CronEmail(['joe@masoniteproject.com'])).send()
)).every('3 days')
```

## New Documentation

This PR adds the task scheduling from the masonite-scheduler repo to the core codebase.

This also adds a new Task Handler class (in keeping with the white paper we set out at the beginning of the project)

This handler class is really just an internal class but we can add tasks to it like we do commands and other things:

```python
self.appllication.make('scheduler').add(Task1(), Task2())
```

There is a wrapper around this if you make a service provider:

```python
from masonite.scheduling import CanSchedule

class ConsoleProvider(Provider, CanSchedule):

    def register(self):
         self.schedule(Task1()).daily_at(13,17) # runs daily at 1pm and 5pm
```

Theres a bunch of diff timing methods in the docs and thats really all there is to this PR ...

### CODE REVIEW

File: src/masonite/scheduling/CanSchedule.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class CanSchedule:
+    def call(self, command):
+        command_class = CommandTask(command)
```
Comment (girardinsamuel) [inline src/masonite/scheduling/CanSchedule.py:3]: Should not this class be imported ?

File: src/masonite/scheduling/Task.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    def every_minute(self):
+        self.run_every = "1 minute"
+        return self
+
+    def every_15_minutes(self):
+        self.run_every = "15 minutes"
+        return self
+
+    def every_30_minutes(self):
+        self.run_every = "15 minutes"
```
Comment (girardinsamuel) [inline src/masonite/scheduling/Task.py:37]:
```suggestion
        self.run_every = "30 minutes"
```

File: src/masonite/scheduling/Task.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        return self
+
+    def every_45_minutes(self):
+        self.run_every = "15 minutes"
```
Comment (girardinsamuel) [inline src/masonite/scheduling/Task.py:41]:
```suggestion
        self.run_every = "45 minutes"
```

File: tests/features/scheduling/test_scheduling.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        time = pendulum.now().on(2018, 1, 1).at(12, 20, 5)
+        self.task.run_every = ""
+        self.task.twice_daily = (1, 13)
+        self.task._date = time
+
+        assert self.task.should_run() is False
+
+    def test_run_at(self):
+        self.task.run_every = ""
+        self.task.run_at = None
```
Comment (girardinsamuel) [inline tests/features/scheduling/test_scheduling.py:130]: Why those two lines to set same attribute ?

File: tests/features/scheduling/test_scheduling.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        self.task.run_at = "13:00"
+
+        time = pendulum.now().on(2018, 1, 1).at(13, 0, 5)
```
Comment (girardinsamuel) [inline tests/features/scheduling/test_scheduling.py:133]: I am not sure of what you're trying to do here. Are some asserts missing ?

File: tests/features/scheduling/test_scheduling.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        time = pendulum.now().on(2018, 1, 1).at(13, 6, 5)
+        self.task._date = time
+
+        assert self.task.should_run() is False
+
+    def test_method_calls(self):
+        task = MockTask()
+        task.at("13:00")
+
+        time = pendulum.now().on(2018, 1, 1).at(13, 0, 5)
```
Comment (girardinsamuel) [inline tests/features/scheduling/test_scheduling.py:152]: What's the goal of two lines here ? There is no call

File: src/masonite/scheduling/Task.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+        pass
+
+    def should_run(self, date=None):
+        """If the task should run"""
+
+        # set the date
+        self._set_date()
+
+        if self._verify_run():
```
Comment (girardinsamuel) [inline src/masonite/scheduling/Task.py:78]: Maybe here you could return `self._verify_run()` directly ?

**Review by girardinsamuel — CHANGES_REQUESTED** [review]
The API looks nice 👍🏼

File: src/masonite/scheduling/Task.py (hunk 7 above)
Comment (josephmancuso) [reply src/masonite/scheduling/Task.py:78]: I like it

File: tests/features/scheduling/test_scheduling.py (hunk 5 above)
Comment (josephmancuso) [reply tests/features/scheduling/test_scheduling.py:133]: just setting the time to 1pm

**Pull request merged**